crates/noirc_evaluator/src/ssa/node.rs

```diff
@@ -0,0 +1,948 @@
+    sle,      //(<=) signed less or equal
+    lt,       //(<) field less
+    gt,       //(>) field greater
+    lte,      //(<=) field less or equal
+    gte,      //(<=) field greater or equal
+    and,
+    not,
+    or,
+    xor,
+    cast,  //convert type
+    ass,   //assignement
+    trunc, //truncate
+
+    //control flow
+    jne, //jump on not equal
+    jeq, //jump on equal
+    jmp, //unconditional jump
+    phi,
+    // todo: call, br,..
+    nop, // no op
+
+    eq_gate, //write a gate enforcing equality of the two sides (to support the constrain statement)
+             //memory todo: load, store, getelementptr?
+}
+
+pub fn is_commutative(op_code: Operation) -> bool {
+    matches!(
+        op_code,
+        Operation::add
+            | Operation::safe_add
+            | Operation::mul
+            | Operation::safe_mul
+            | Operation::and
+            | Operation::or
+            | Operation::xor
+    )
+}
+
+pub fn is_binary(op_code: Operation) -> bool {
+    match op_code {
+        Operation::add => true,
+        Operation::safe_add => true,
+        Operation::sub => true,
+        Operation::safe_sub => true,
+        Operation::mul => true,
+        Operation::safe_mul => true,
+        Operation::udiv => true, //(/) unsigned division
+        Operation::sdiv => true,
+        Operation::urem => true,
+        Operation::srem => true,
+        Operation::fmod => true,
+        Operation::fneg => true,
+        Operation::fdiv => true,
+        Operation::div => true,
+        Operation::eq => true,
+        Operation::ne => true,
+        Operation::ugt => true,
+        Operation::uge => true,
+        Operation::ult => true,
+        Operation::ule => true,
+        Operation::sgt => true,
+        Operation::sge => true,
+        Operation::slt => true,
+        Operation::sle => true,
+        Operation::lt => true,
+        Operation::gt => true,
+        Operation::lte => true,
+        Operation::gte => true,
+        Operation::and => true,
+        Operation::not => false,
+        Operation::or => true,
+        Operation::xor => true,
+        Operation::cast => false,
+        Operation::ass => false,
+        Operation::trunc => true,
+        Operation::jne | Operation::jeq | Operation::jmp => false,
+        Operation::phi => false,
+        Operation::nop => false,
+        Operation::eq_gate => true,
+    }
+}
+
+pub fn to_operation(op_kind: HirBinaryOpKind, op_type: ObjectType) -> Operation {
+    match op_kind {
+        HirBinaryOpKind::Add => Operation::add,
+        HirBinaryOpKind::Subtract => Operation::sub,
+        HirBinaryOpKind::Multiply => Operation::mul,
+        HirBinaryOpKind::Equal => Operation::eq,
+        HirBinaryOpKind::NotEqual => Operation::ne,
+        HirBinaryOpKind::And => Operation::and,
+        HirBinaryOpKind::Or => Operation::or,
+        HirBinaryOpKind::Xor => Operation::xor,
+        HirBinaryOpKind::Divide => {
+            if op_type.is_signed() {
```

Comment:
Can you create an enum that encapsulates the Signed, Field and Unsigned variants, then convert the ObjectType to this and match on that new variant?

This way we can get rid of these if statements for an exhaustive match

Reply:
Done